lib/gnat.ex

```diff
@@ -141,15 +135,19 @@ defmodule Gnat do
     sub = Command.build(:sub, request.inbox, sid, [])
     unsub = Command.build(:unsub, sid, [max_messages: 1])
     pub = Command.build(:pub, request.topic, request.body, reply_to: request.inbox)
-    receivers = Map.put(state.receivers, sid, request.recipient)
-    next_sid = sid + 1
     :ok = :gen_tcp.send(state.tcp, [sub, unsub, pub])
-    {:reply, {:ok, sid}, %{state | receivers: receivers, next_sid: next_sid}}
+    state = add_subscription_to_state(state, sid, request.recipient) |> cleanup_subscription_from_state(sid, max_messages: 1)
+    next_sid = sid + 1
+    {:reply, {:ok, sid}, %{state | next_sid: next_sid}}
   end
-  def handle_call({:unsub, sid, opts}, _from, state) do
-    command = Command.build(:unsub, sid, opts)
-    :ok = :gen_tcp.send(state.tcp, command)
-    {:reply, :ok, state}
+  def handle_call({:unsub, sid, opts}, _from, %{receivers: receivers}=state) do
+    case Map.has_key?(receivers, sid) do
+      false -> {:reply, {:error, :no_such_subscription}, state}
```

Comment:
I would still consider this a success if the caller is cautiously cleaning up a subscription that auto unsub'd, like we do when we perform a double response request. I like that you're distinguishing between what happens, but returning an `error` might be the wrong way to describe it.